### DISCUSSION TITLE
cookies vs localStorage vs session storage

### DISCUSSION THREAD

**Discussion by jswith** [woowacourse-study/2022-woowahan-bansanghwe#104] [Network]
## Web Storage

웹 스토리지란 무엇일까? HTML5가 등장하기 이전에는 애플리케이션의 데이터를 4KB 남짓에 불과한 Cookies에 저장해야 했고, Cookies는 매 서버 요청마다 함께 전송되었다. 이는 웹 사이트의 성능에 영향을 줄 수 있을 뿐만 아니라 보안 상의 이슈도 존재했다. 하지만 웹 스토리지의 등장 이후 웹 사이트 성능에 영향을 주지 않고 많은 양의 데이터를 로컬에 저장할 수 있게 되었다. 즉 웹 애플리케이션은 웹 스토리지를 이용하여 사용자의 브라우저 내에 로컬 데이터를 저장할 수 있게 된 것이다.

웹 스토리지는 Local Storage와 Session Storage가 있다. 이 둘은 Cookies보다 저장소의 용량이 훨씬 크며, 정보가 서버에 전송되지 않는다. Local Storage와 Session Storage는 거의 유사하게 동작하지만 몇 가지 차이점이 존재한다. 먼저, Local Storage와 Session Storage에 대해서 알아보자.

## Local Storage와 Session Storage

Local Storage와 Session Storage의 공통점과 차이점, 그리고 Local Storage와 Session Storage의 사용법에 대해 알아본다.

### 공통점

Local Storage와 Session Storage는 UTF-16 string 형식으로 데이터를 저장하며, 문자마다 2바이트를 사용한다. 자바스크립트의 객체와 마찬가지로 숫자 형태의 키는 자동적으로 string으로 변환된다. Local Storage와 Session Storage는 Cookies와 다르게 서버로 데이터가 전송되지 않는다. 또한 저장소의 용량이 5MB로 4KB의 크기를 가진 Cookies보다 훨씬 크다. Local Storage 및 Session Storage에 저자오딘 데이터는 클라이언트 내에서 쉽게 읽거나 조작할 수 있으므로 중요 데이터나 보안 관련 데이터를 데이터를 저장하는 것은 피해야 한다.

### 차이점

Session Storage는 페이지 세션이 열려있는 동안, 즉 브라우저의 탭이 닫히기 전까지는 데이터를 유지한다. 따라서 탭을 닫지 않는 경우에는 페이지를 다시 로드하더라도 데이터가 유지된다. 하지만 브라우저의 탭을 닫는 순간 저장된 데이터가 사라지게 된다. 반면 Local Storage는 사용자가 브라우저 데이터를 초기화하거나 Local Storage의 데이터를 삭제하기 전까지는 브라우저를 닫거나 탭을 닫더라도 기존의 데이터를 유지한다.

Local Storage의 데이터는 기본적으로 도메인 단위로 공유되지만, 프로토콜이 다를 경우에는 도메인이 같아도 공유되지 않는다. 즉 origin 단위(per domain and protocol)로 공유된다. 예를 들면 `http://example.com`과 `https://example.com`은 서로 다른 Local Storage 객체를 사용한다. Session Storage는 동일한 도메인의 창이나 탭을 여러 개 띄울 경우 전부 다른 Session Storage 객체를 사용한다. 탭이나 창을 새로 여는 순간 최상위 브라우징 컨텍스트 값을 가진 세션이 생성되기 때문이다.

> Browsing Context란?
>
> A browsing context is an environment in which a browser displays a Document. In modern browsers, **it usually is a tab**, but can be a window or even only parts of a page, like a frame or an iframe.
>
> Each browsing context has an origin (that of the active document) and an ordered history of previously displayed documents.
>
> Communication between browsing contexts is severely constrained. Between browsing contexts of the same origin, a BroadcastChannel can be opened and used.

### 사용법

```javascript
// Local Storage
localStorage.setItem("myCat", "Tom"); // 데이터 저장

const cat = localStorage.getItem("myCat"); // 데이터 조회

localStorage.removeItem("myCat"); // 데이터 삭제

localStorage.clear(); // 초기화

// Session Storage
sessionStorage.setItem("key", "value"); // 데이터 저장

let data = sessionStorage.getItem("key"); // 데이터 조회

sessionStorage.removeItem("key"); // 데이터 삭제

sessionStorage.clear(); // 초기화
```

## Cookies

Cookie는 주로 사용자 인증, 세션 및 광고 추적을 위한 식별 정보을 저장하는 데 사용한다. 동일한 도메인에 대한 요청이 있을 때마다, 페이지에 유효한 모든 Cookie는 브라우저에서 서버로 전송된다.

Local Storage와 Session Storage와 마찬가지로 Cookies 역시 사용자가 조작할 수 있으며, 데이터를 평문으로 읽을 수 있다. 또한 SSL을 사용하지 않을 경우, 전송 중에 Cookie 정보가 탈취될 수 있는 위험도 존재한다. Cookie 정보를 탈취한 사용자는 Cookie의 복사본을 통해 웹 애플리케이션에 마치 해당 사용자인 것처럼 행동할 수 있다.

> **SSL?**
>
> SSL(Secure Sockets Layer)은 [암호화](https://www.cloudflare.com/learning/ssl/what-is-encryption/) 기반 인터넷 보안 [프로토콜](https://www.cloudflare.com/learning/network-layer/what-is-a-protocol/)입니다. 인터넷 통신의 개인정보 보호, 인증, 데이터 무결성을 보장하기 위해 Netscape가 1995년 처음으로 개발했습니다. SSL은 현재 사용 중인 [TLS](https://www.cloudflare.com/learning/ssl/transport-layer-security-tls/) 암호화의 전신입니다.
>
> SSL은 높은 수준의 [개인정보 보호](https://www.cloudflare.com/learning/privacy/what-is-data-privacy/)를 제공하기 위해, 웹에서 전송되는 데이터를 암호화합니다. 따라서, 데이터를 가로채려는 자는 거의 해독할 수 없는 복잡한 문자만 보게 됩니다.

물론 Cookie의 보안상 취약점을 보완할 수 있는 방법도 존재한다. Cookie의 HTTP only flag를 사용하여 JavaScript를 통한 Cookie 접근을 막으면, Cross-Site Scripting(XSS)이나 Script Injection과 같은 위험으로부터 어느 정도 보호받을 수 있다.

> **Cross-Site Scripting(XSS)?**
>
> 크로스 사이트 스크립팅(Cross Site Scripting, XSS)은 공격자가 상대방의 브라우저에 스크립트가 실행되도록 해 사용자의 세션을 가로채거나, 웹사이트를 변조하거나, 악의적 콘텐츠를 삽입하거나, 피싱 공격을 진행하는 것을 말합니다. XSS 공격은 스크립트 언어와 취약한 코드를 공격 대상으로 합니다

> **Script Injection?**
>
> 브라우저에서 자바스크립트를 삽입시키는 공격이다. 브라우저에서 제공하는 Console을 통해 조작 가능하다. 만약 Client-Side에 민감한 데이터를 넣어놨다면 해당 공격을 통해 탈취가 가능하다.

## 그림으로 비교하기

![image](https://user-images.githubusercontent.com/64275588/193437926-2b214c5f-ceb7-45ba-8b60-de4bc755b796.png)

## References

- https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API
- https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage
- https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage
- https://developer.mozilla.org/en-US/docs/Glossary/Browsing_context#:~:text=A%20browsing%20context%20is%20an,history%20of%20previously%20displayed%20documents.
- https://stackoverflow.com/questions/19867599/what-is-the-difference-between-localstorage-sessionstorage-session-and-cookies
- https://www.cloudflare.com/ko-kr/learning/ssl/what-is-ssl/
- https://nordvpn.com/ko/blog/xss-attack/
- https://blog.cobalt.run/posts/basic-web-hacking

**Comment by airman5573** [comment]
언제나 상세하고 친절한 민초의 설명 👍

**Comment by al-bur** [comment]
간단히 자기소개 해주세요

**Comment by al-bur** [comment]
웹 어플리케이션에서 메모리가 아닌 곳에 저장하려면 어디에 저장할 수 있나요?

**Comment by al-bur** [comment]
web storage는 뭔가요?

**Comment by al-bur** [comment]
왜 webstorage가 등장했나요?

**Comment by al-bur** [comment]
왜 storage가 보안에 안전한가요?

**Comment by al-bur** [comment]
header에 노출 되는 게 보안적으로 안좋은건가요?

**Comment by al-bur** [comment]
cookie에 보안을 높이려면 어떻게 할까요?

**Reply by al-bur** [reply]
http only, secure

**Comment by al-bur** [comment]
xss는 어떤 건가요?

**Comment by al-bur** [comment]
다른 탭(도메인)에서도 쿠키가 공유가 되는 건가요?

**Reply by al-bur** [reply]
체크 필요

**Comment by greenblues1190** [comment]
프로젝트에서 쿠키를 사용해본 경험이 있나요?

**Comment by greenblues1190** [comment]
쿠키로 리프레시 토큰 인증 인가를 진행한 프로세스를 자세하게 설명해주실 수 있나요?

**Comment by al-bur** [comment]
왜 리프레시 토큰을 사용하게 됐나요?

**Reply by al-bur** [reply]
리프레시 토큰이 30분??

**Comment by greenblues1190** [comment]
토큰을 이용해서 '기억하기' 기능을 구현한다면 어떻게 구현해야할까요?

**Comment by al-bur** [comment]
tcp 연결 지속성에 대한 시간 기준은?

**Comment by greenblues1190** [comment]
쿠키는 패킷의 어디에 저장이 되나요?

**Comment by al-bur** [comment]
HTTPS는 body만 암호화를 해주나?

**Comment by airman5573** [comment]
cookie를 사용했을때 어떤 보안상의 이슈가 생길 수 있나요?

**Comment by airman5573** [comment]
cookie를 사용해 본 경험이 있나요?

**Comment by airman5573** [comment]
cookie가 탄생한 배경에 대해 소개해 주세요

**Comment by airman5573** [comment]
refresh token을 왜 사용하셨나요?

**Comment by airman5573** [comment]
local storage를 사용했을때 발생할 수 있는 보안적인 이슈에 대해 소개해주세요

**Comment by airman5573** [comment]
XSS 공격은 어떤 경로로 이뤄지나요?
